### PR TITLE
topology1: add support for cs42l42-sdw

### PR DESCRIPTION
**Pull request by davidrau-renesas-opensource**
Adding support for cs42l42 via SoundWire

### CODE REVIEW

**Comment by sofci** [comment]
Can one of the admins verify this patch?
> reply **test&nbsp;this&nbsp;please** to run this test once

**Comment by davidrau-renesas-opensource** [comment]
test this please

**Comment by lgirdwood** [comment]
test this please

File: tools/topology/topology1/sof-adl-sdw-cs42l42.m4 [hunk 1]
```diff
@@ -0,0 +1,286 @@
+#
+# Topology for Tigerlake with rt711 + rt1308 (x2).
```
Comment (kv2019i) [inline tools/topology/topology1/sof-adl-sdw-cs42l42.m4:2]: This comment needs to be updated.

**Review by kv2019i — CHANGES_REQUESTED** [review]
Thanks you @davidrau-renesas-opensource for the submission. One minor note, but a bigger ask is that can you submit this against the stable-v2.2 branch (make a pull-request to this branch). We have dropped topology1 support for Intel platforms from SOF mainline, and support for topology1 TGL/ADL platforms is now done in the stable-v2.2 branch.

File: tools/topology/topology1/sof-adl-sdw-cs42l42.m4 [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+#
+
```
Comment (plbossart) [inline tools/topology/topology1/sof-adl-sdw-cs42l42.m4:4]: why is there a need to create a new topology just to support a new headset codec?

Put differently, why not use the existing topology or describe what's missing?

Topology files are a PITA to maintain, less is more really.

File: tools/topology/topology1/sof-adl-sdw-cs42l42.m4 [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+# playback DAI is iDisp3 using 2 periods
+# Buffers use s32le format, 1000us deadline with priority 0 on core 0
+DAI_ADD(sof/pipe-dai-playback.m4,
+	8, HDA, 2, iDisp3,
+	PIPELINE_SOURCE_8, 2, s32le,
+	1000, 0, 0, SCHEDULE_TIME_DOMAIN_TIMER)
+
+# PCM Low Latency, id 0
+dnl PCM_PLAYBACK_ADD(name, pcm_id, playback)
+PCM_PLAYBACK_ADD(CS42L42 Jack Out, 0, PIPELINE_PCM_30)
```
Comment (plbossart) [inline tools/topology/topology1/sof-adl-sdw-cs42l42.m4:244]: why redefine names? What was wrong with just 'Jack Out' which is used by all existing solutions?

**Review by plbossart — CHANGES_REQUESTED** [review]
Blocking with "request changes" until we agree about what's missing in existing topologies first.

File: tools/topology/topology1/sof-adl-sdw-cs42l42.m4 (hunk 2 above)
Comment (davidrau-renesas-opensource) [reply tools/topology/topology1/sof-adl-sdw-cs42l42.m4:4]: Thanks for the suggestions.
These modifications I did is a temproary solution to enable CS42L42 codec in ADL platform.

> Topology files are a PITA to maintain, less is more really.
Agreed with you. :)

File: tools/topology/topology1/sof-adl-sdw-cs42l42.m4 (hunk 3 above)
Comment (davidrau-renesas-opensource) [reply tools/topology/topology1/sof-adl-sdw-cs42l42.m4:244]: Understood.

**Comment by plbossart** [comment]
@davidrau-renesas-opensource it wouldn't hurt if you introduced a bit more the problem you are trying to solve and in which environment. We are not really aware of any ADL platform based on CS42L42?

**Comment by davidrau0914** [comment]
> @davidrau-renesas-opensource it wouldn't hurt if you introduced a bit more the problem you are trying to solve and in which environment. We are not really aware of any ADL platform based on CS42L42?

Here is the story.
Previously, I purchased a development board (Alder-Lake) to study how SoundWire interface work with the audio codec.
And the onboard codec is CS42L42-SDW that not supported by SOF as before.
I reached out SOF community and got some feedbacks from the maintainer about how to add ACPI table, modify SOF driver and tplg file.

After such modificatins, I can verify the playback function and then push this tplg file that the maintainer asked.

**Comment by davidrau-renesas-opensource** [comment]
> @davidrau-renesas-opensource it wouldn't hurt if you introduced a bit more the problem you are trying to solve and in which environment. We are not really aware of any ADL platform based on CS42L42?

Here is the story.
Previously, I purchased a development board (Alder-Lake) to study how SoundWire interface work with the audio codec.
And the onboard codec is CS42L42-SDW that not supported by SOF as before.
I reached out SOF community and got support from the maintainer about how to add ACPI table, modify SOF driver and tplg file.

After such modificatins, I can verify the playback function and then push this tplg file that the maintainer asked.

**Comment by lgirdwood** [comment]
Glad its working for you now @davidrau-renesas-opensource, did you say what SOF version you are using? Is it v2.2 with IPC3 or v2.3+ with IPC4 ? as this determines what we need to upstream as some features are being deprecated. Thanks

**Comment by davidrau-renesas-opensource** [comment]
> Glad its working for you now @davidrau-renesas-opensource, did you say what SOF version you are using? Is it v2.2 with IPC3 or v2.3+ with IPC4 ? as this determines what we need to upstream as some features are being deprecated. Thanks

@lgirdwood 
Thanks for the kind support always.
 According to the system log in my dev-kit, the SOF mechanism bases on **IPC3**
`
[   16.171548] snd_sof:sof_ipc3_fw_parse_ext_man: sof-audio-pci-intel-tgl 0000:00:1f.3: found sof_ext_man header type 1 size 0x1a0`
`[   16.171550] snd_sof:sof_ipc3_fw_parse_ext_man: sof-audio-pci-intel-tgl 0000:00:1f.3: found sof_ext_man header type 6 size 0x20`
`[   16.171552] snd_sof:sof_ipc3_fw_parse_ext_man: sof-audio-pci-intel-tgl 0000:00:1f.3: found sof_ext_man header type 0 size 0x50`
`[   16.171556] snd_sof:sof_ipc3_fw_parse_ext_man: sof-audio-pci-intel-tgl 0000:00:1f.3: found sof_ext_man header type 2 size 0x70`
`[   16.171557] snd_sof:sof_ipc3_get_cc_info: sof-audio-pci-intel-tgl 0000:00:1f.3: Firmware info: used compiler XCC 12:0:8 <RG-2017.8-linux> used optimization flags -O2`

And I clone SOF repository from your Github to create this tplg source file for CS42L42.
The version info of SOF repository is "**2.7.99**" (main branch)

Thanks.

**Comment by davidrau-renesas-opensource** [comment]
> @davidrau-renesas-opensource Can you submit a pull req with this commit o stable-v2.2 branch? It has the latest tools/topology1 so this PR should apply cleanly on top of it.

@kv2019i Thanks for your suggestions. :)
I will switch to stable-v2.2 branch and pull req with my commit as soon.

**Pull request closed without merging**